btree/btree.c

```diff
@@ -4016,313 +4130,733 @@ static struct vkvv_head *vkvv_data(const struct nd *node)
+				 rec[index].key_offset;
+		total_vsize    = rec[count].val_offset -
+				 rec[index].val_offset;
+
+		while (t_count >= index) {
+			rec[t_count+1].key_offset = rec[t_count].key_offset;
+			rec[t_count+1].val_offset = rec[t_count].val_offset;
+
+			rec[t_count+1].key_offset = rec[t_count + 1].key_offset
+						    + ksize;
+			rec[t_count+1].val_offset = rec[t_count + 1].val_offset
+						    + vsize;
+			t_count --;
+		}
+
+		memmove(start_key_addr + ksize, start_key_addr, total_ksize);
+		memmove(start_val_addr - total_vsize - vsize, start_val_addr -
+			total_vsize, total_vsize);
+	}
+}
+
+
+static void vkvv_inode_make(struct slot *slot)
+{
+	int               index          = slot->s_idx;
+	uint32_t          ksize          =
+			m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
+	struct vkvv_head *h              = vkvv_data(slot->s_node);
+	uint16_t          count          = h->vkvv_used;
+	uint32_t          vspace         = vkvv_get_vspace();
+	uint32_t          vsize          = vspace;
+	uint32_t         *count_offset;
+	uint32_t         *index_offset;
+	uint32_t         *index_offset_2;
+	uint32_t         *t_offset;
+	uint32_t         *t_offset_2;
+	uint32_t          total_ksize;
+	uint32_t          total_vsize;
+	uint32_t         *offset;
+	int               t_count;
+	void             *start_key_addr;
+	void             *start_val_addr;
+
+	if (index == count) {
+		/**
+		 * No need to do memmove() as data will be added to the end of
+		 * current series of keys and values. Just update the
+		 * directory to keep a record of the next possible offset.
+		 */
+
+		index_offset = vkvv_get_key_offset(slot->s_node, index);
+
+		if (index == 0)
+			*index_offset = ksize;
+		else {
+			 index_offset_2 = vkvv_get_key_offset(slot->s_node,
+							      index - 1);
+			*index_offset   = *index_offset_2 + ksize;
+		}
+	} else {
+		start_key_addr = vkvv_key(slot->s_node, index);
+		start_val_addr = vkvv_val(slot->s_node, index);
+		total_vsize    = vspace *(count - index);
+		t_count        = count;
+
+		if (index == 0) {
+			index_offset = vkvv_get_key_offset(slot->s_node,
+							   count - 1);
+			total_ksize  = *index_offset;
+		} else {
+			count_offset = vkvv_get_key_offset(slot->s_node,
+							   count - 1);
+			index_offset = vkvv_get_key_offset(slot->s_node,
+							   index - 1);
+			total_ksize  = *count_offset - *index_offset;
+		}
+
+		start_val_addr -= INT_OFFSET;
+
+		memmove(start_key_addr + ksize, start_key_addr, total_ksize);
+		memmove(start_val_addr - total_vsize - vsize, start_val_addr -
+			total_vsize, total_vsize);
+
+		while (t_count > index) {
+			offset  = vkvv_get_key_offset(slot->s_node, t_count);
+			*offset = *offset + ksize;
+			t_count --;
+		}
+
+		if (index == 0) {
+			t_offset  = vkvv_get_key_offset(slot->s_node, t_count);
+			*t_offset = ksize;
+		} else {
+			t_offset   = vkvv_get_key_offset(slot->s_node, t_count);
+			t_offset_2 = vkvv_get_key_offset(slot->s_node, t_count
+							 - 1);
```

Comment:
instead of writing -1 on next like take whole 2nd argument on next line t_count -1

Reply:
Done